### PR TITLE
ETH-206: Add test

### PR DESCRIPTION
**Pull request by jtakalai**
also fix broken permission object handling

### CODE REVIEW

File: src/CoreAPIClient.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export default class CoreAPIClient {
         return this._get(apiUrl)
     }
 
-    private async _post(apiUrl: string, body: any): Promise<Response> {
+    private async _post(apiUrl: string, body: any, method = "POST"): Promise<Response> {
```
Comment (kare) [inline src/CoreAPIClient.ts:64]: When refactoring code default parameter values are usually troublesome. What's the reason for this change?

File: test/integration.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        log(process.execPath)
+        watcherProcess = spawn(process.execPath, [watcherExecutable], { env: watcherEnv })
+        if (!watcherProcess.stdout || !watcherProcess.stderr) { throw new Error("No stdout/stderr in spawned process") }
+        watcherProcess.stdout.on("data", (data) => { log(`(server stdout) ${data.toString().trim()}`) })
+        watcherProcess.stderr.on("data", (data) => { log(`(server stderr) ${data.toString().trim()}`) })
+        watcherProcess.on("close", (code) => { log(`server exited with code ${code}`) })
+        watcherProcess.on("error", (err) => { log(`server ERROR: ${err}`) })
+
+        // wait until server is started, also get url from output
+        await untilStreamMatches(watcherProcess.stdout, /Playback done. Starting watcher.../)
```
Comment (kare) [inline test/integration.ts:206]: Reading process stdout is probably fine, but it's also brittle. Would it be possible to have some event functions in the watcher that you register to listen and do a callback function. Ie. you get notified by the watcher that this thing just happened and your callback function gets called.

File: test/integration.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    CoreAPIClient.DEFAULT_FETCH_FUNC,
+    CoreAPIClient.DEFAULT_GET_SESSION_TOKEN_FUNC,
+    adminKey
+)
+const trustedApiClient = new CoreAPIClient(
+    STREAMR_API_URL,
+    CoreAPIClient.DEFAULT_FETCH_FUNC,
+    CoreAPIClient.DEFAULT_GET_SESSION_TOKEN_FUNC,
+    DEVOPS_KEY
+)
```
Comment (kare) [inline test/integration.ts:91]: I'd move these variables closer to where they are used and I'd narrow the scope. It's a bit unnecessary to scroll up and down to see where variables are declared.

File: src/CoreAPIClient.ts (hunk 1 above)
Comment (jtakalai) [reply src/CoreAPIClient.ts:64]: well, it's the same as POST otherwise, so that keeps the code a bit more DRY I guess.

File: test/integration.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
+import { spawn } from "child_process"
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:1]: Why can't CI spawn processes?

File: test/integration.ts [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        function check(buffer: Buffer) {
+            const data = buffer.toString()
+            const match = data.match(regex)
+            if (match) {
+                if (stream.off) { stream.off("data", check) }
+                resolve(match)
+            }
+        }
+        stream.on("data", check)
+    })
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:110]: It will resolve when there's a matching item in the log. Until then it will be pending. The quote does not say that a Promise ever needs to reject.

File: test/integration.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+    let productId: string
+    let productIdBytes: string
+    before(async function () {
+        this.timeout(30000)
+
+        // fund the devops key
+        const fundTx1 = await wallet.sendTransaction({
+            to: watcherWallet.address,
+            value: parseEther("10")
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:123]: it would be possible, but not more readable. That's why IDEs have "jump to definition".

File: test/integration.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+        // add stream to product
+        const streamIdEncoded = encodeURIComponent(streamId)
+        let status = 0
+        do {
+            // @ts-ignore-next-line private
+            const addRes = await ownerApiClient._post(`${STREAMR_API_URL}/products/${productId}/streams/${streamIdEncoded}`, {}, "PUT")
+            log("Add stream to product returned %s %s", addRes.status, addRes.statusText)
+            status = addRes.status
+        } while (status !== 204)
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:161]: that was a hack to work around web3j timing out on the server. There is no reason it should return 4xx or 5xx here.

File: test/integration.ts (hunk 2 above)
Comment (jtakalai) [reply test/integration.ts:206]: not sure how that works over processes...

The point of starting a separate process is, I guess, to test it as it would be running on real server: a separate process. Of course creating a watcher etc. in the test would make it more poke-able but also less realistic.

Starting a separate process is a straightforward way to test running the js file as a script. I guess you can run it in the same process too (`require` does something like that I guess?).

File: test/integration.ts (hunk 3 above)
Comment (jtakalai) [reply test/integration.ts:91]: moving them as close as possible would be a bit... misleading in the sense that there is currently only one test. Stuff might need to be moved out after adding another test. I think more generally the way our tests are laid our and formatted could be a wider discussion. I wouldn't change this now.

File: src/CoreAPIClient.ts (hunk 1 above)
Comment (jtakalai) [reply src/CoreAPIClient.ts:64]: If I refactor the test so that I don't need a PUT, I can revert that change for sure.

Comment (kare) [reply src/CoreAPIClient.ts:64]: My point is: Never use default values for function parameters, they cause trouble later on.

File: test/integration.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:1]: Comment is unrelated to line number one. Just a remark that currently CI (Continuous Integration) does not execute integration tests. However, integration tests must be executed as part of the CI cycle.

File: test/integration.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+    },
+    streamr: {
+        rpcEndpoints: [{
+            url: MATIC_SERVER_URL,
+        }],
+        contracts: {
+            "StreamRegistry": STREAM_REGISTRY_ADDRESS,
+        }
+    }
+} = Chains.load("development")
```
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:40]: Where does the value of variable `STREAM_REGISTRY_ADDRESS` come from? It seems to me that it doesn't come from config package?

File: test/integration.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+import CoreAPIClient from "../src/CoreAPIClient"
+
+const { log } = console
+
+import { Chains } from "@streamr/config"
+import assert from "assert"
+
+const {
+    ethereum: {
```
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:22]: Fair enough, being used to strongly typed languages it just looks to me like the type information is lost there. I guess you can cast `ethereum` back to `config.Chain` if you like?

File: test/integration.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:108]: Aah I see, but `Promise` constructor takes two arguments which are non-optional and the other one is `reject`. To me it looks like it's missing?

File: test/integration.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:159]: What if 204 never happens?

File: test/integration.ts (hunk 2 above)
Comment (kare) [reply test/integration.ts:204]: That's just an idea. I would still consider it an integration test if it would be something like (pseudo):
```
const watcher = new Watcher(function onServerStarted() {
    runIntegrationTest()
})
```

File: test/integration.ts (hunk 3 above)
Comment (kare) [reply test/integration.ts]: If it is used only once then the rationale is even heavier. Are you absolutely sure there is going to be another test that uses all of those variables?

File: test/integration.ts (hunk 2 above)
Comment (kare) [reply test/integration.ts:204]: I'd call it an integration test even if it disregards things like command line arguments or doesn't read environment variables in `main` function.

File: src/CoreAPIClient.ts (hunk 1 above)
Comment (jtakalai) [reply src/CoreAPIClient.ts:64]: ah well, in this case it saves the duplication of the complete function body. I don't disagree with that in general.

File: test/integration.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:40]: that's the new name the value is bound to. It looks like that because it's also the "env variable" passed as env to the spawned watcher.

File: test/integration.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:108]: you're right, it's not used, so it can be omitted. Yay javascript!

File: test/integration.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:121]: `parseEther` is a utility that someone might write their own, sure. But let's not do that.

File: test/integration.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:159]: then the test will fail with timeout. It should happen every time, but doesn't because of that mystery timeout (that I'm totally not going to debug now :D )

File: test/integration.ts (hunk 2 above)
Comment (jtakalai) [reply test/integration.ts:204]: sure, but spawning that process isn't a big deal IMO. Not as pretty as something like perl maybe but still okay.

File: test/integration.ts (hunk 3 above)
Comment (jtakalai) [reply test/integration.ts]: it's more like: if there are many test cases, will all of them use these global variables? Where should we put them? So far I've put them outside the describe. They could be inside. Should we even require a describe? Just `it` works too. I don't think simple "put it close to the usage" is the best here, if the result is that every test looks different.

File: test/integration.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:22]: Config package is version 0.0.3 at the moment. We can modify config package's API as we see fit. I'd like to hear your input!

[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:22]: well I'll let my input heard once I have some :)  I think it's kind of hard to design an ergonomic config thing in vacuum, only after trying it out we can iterate.

File: test/integration.ts (hunk 3 above)
Comment (kare) [reply test/integration.ts]: That's just speculation. Will there ever be another test? Write code as the current situation is/requires and don't speculate about unknown future, right?

Comment (jtakalai) [reply test/integration.ts]: There will be other tests in other files, and it will be good for readability if they're all formatted in the same way. This is my formatting for now, until further general discussion on how tests should be laid out.

File: src/CoreAPIClient.ts (hunk 1 above)
Comment (kare) [reply src/CoreAPIClient.ts:64]: Using default parameters don't save a complete function body.

File: test/integration.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:22]: If the internal presentation of the config packages changes significantly. These lines that are destructured need to be rewritten or not?

File: test/integration.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:40]: What new name you are referring to?

File: test/integration.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kare) [reply test/integration.ts:121]: I'd say package is relevant unless it comes from Node standard library API. Package is very significant if it comes from a dependency.

File: .github/workflows/ci-cd.yaml [hunk 10]
```diff
@@ -18,11 +18,21 @@ jobs:
...
+          npm run build
+          npm run test
+      - name: Start Streamr Docker Stack
+        uses: streamr-dev/streamr-docker-dev-action@v1.0.0
+        with:
+          services-to-start: 'parity-sidechain-node0 parity-node0 core-api nginx'
+      - name: Integration Test
+        env:
+          DEBUG: '*'
+          TEST_TIMEOUT: '600000'
```
[the remark this answers is not in the document]
Comment (jtakalai) [reply .github/workflows/ci-cd.yaml:34]: what should it be? I'll set it to `development`.

[the remark this answers is not in the document]
Comment (kare) [reply .github/workflows/ci-cd.yaml:34]: 👍

**Review by kare — APPROVED** [review]

File: test/integration.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jtakalai) [reply test/integration.ts:40]: variable binding. That line is equivalent to
```
const STREAM_REGISTRY_ADDRESS = Chains.load("development").streamr.contracts.StreamRegistry
```
